<a href="https://colab.research.google.com/github/AbdulWasay65/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AbdulWasay65/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why
Lane: Refresh / Content Opportunity Scoring

I want to investigate how observable search-performance signals can help prioritize which pages should receive content-refresh review first. I chose this lane because the starter dataset shows that simple assumptions do not always explain page performance. Search volume had almost no correlation with 90-day impressions (0.001), while average position and CTR showed a weak negative relationship (-0.073). The starter pipeline also showed a difference between a hand-written baseline and a learned ranking model, making content-refresh prioritization a useful decision-support problem to investigate further.

In [ ]:
# This cell is for CODE
import os
import sys
import subprocess

REPO_URL = "https://github.com/AbdulWasay65/flyrank-ml-internship"
REPO_DIR = "/content/flyrank-ml-internship"

# Clone your GitHub repository if it isn't already in this runtime
if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

os.chdir(REPO_DIR)

print("Working directory:", os.getcwd())
print("Repo exists:", os.path.isdir(REPO_DIR))



# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Working directory: /content/flyrank-ml-internship
Repo exists: True


In [ ]:
# This cell is for CODE
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print(f"Rows: {len(df):,}")

corr_volume = df["search_volume"].corr(df["impressions_90d"])
corr_position = df["avg_position"].corr(df["ctr"])

print(f"search_volume vs impressions_90d correlation: {corr_volume:.3f}")
print(f"avg_position vs CTR correlation: {corr_position:.3f}")
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Rows: 30,000
search_volume vs impressions_90d correlation: 0.001
avg_position vs CTR correlation: -0.073


## 2. The question: decision, action, cost of a wrong call

The decision is which pages should be prioritized for content-refresh review. The main user of the output would be an SEO or content team deciding where to spend limited review and optimization time. The output would be a ranked list of pages with signals indicating which pages may deserve attention first. A reviewer could then inspect the highest-ranked pages and decide whether a refresh is appropriate. A wrong recommendation has an opportunity cost: a team could spend time refreshing a page that does not need it, while a page that would benefit from review may be missed or delayed. ML can help by combining several measurable signals consistently rather than relying only on a single hand-written rule.

In [ ]:
# This cell is for CODE
print(f"Pages available for prioritization: {len(df):,}")
print(f"Declining pages: {(df['trend_direction'] == 'down').sum():,}")
print(f"Declining rate: {(df['trend_direction'] == 'down').mean():.3f}")
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Pages available for prioritization: 30,000
Declining pages: 16,262
Declining rate: 0.542


## 3. Quick look at the data (2-3 real numbers)

The starter dataset contains 30,000 pages, and 54.2% of the pages are labeled as declining. This suggests that there is a substantial set of pages where prioritization could be useful. Search volume and 90-day impressions have a correlation of only 0.001 in the starter data, which suggests that search volume alone is not a strong indicator of the impressions a page actually receives. Together, these observations make it worth investigating whether multiple measurable signals can provide better decision support for content-refresh prioritization than relying on a single simple assumption.

In [ ]:
# This cell is for CODE
print(f"Total pages: {len(df):,}")

declining_rate = (df["trend_direction"] == "down").mean()
print(f"Declining page rate: {declining_rate:.3f}")

search_volume_impressions_corr = df["search_volume"].corr(
    df["impressions_90d"]
)
print(
    f"search_volume vs impressions_90d correlation: "
    f"{search_volume_impressions_corr:.3f}"
)
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Total pages: 30,000
Declining page rate: 0.542
search_volume vs impressions_90d correlation: 0.001


## 4. Careful words: what I can and can't claim

I will be able to report observed patterns in the available search-performance data and use them as directional evidence for prioritizing pages for content-refresh review. The eventual output should be treated as decision support: it can help a content or SEO team decide which pages deserve investigation first, but it does not replace human review.

I will not claim that the analysis proves why a page is declining, proves causation, or predicts Google's ranking algorithm. Any model results will be interpreted as measured performance on the available data and validation setup, not as proof of how search engines work.

In [ ]:
# This cell is for CODE
print("Careful-claim check:")
print(f"Dataset contains {len(df):,} page-level observations.")
print("Results will be treated as observed and directional, not causal.")
print("The output is intended for decision support, not prediction of Google's algorithm.")
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Careful-claim check:
Dataset contains 30,000 page-level observations.
Results will be treated as observed and directional, not causal.
The output is intended for decision support, not prediction of Google's algorithm.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.